libeos-payg/fake-clock.c

```diff
@@ -110,15 +110,20 @@ typedef struct {
   GSource parent;
 
   EpgFakeClock *clock; /* owned */
-  gint64 ready_time_secs;
+  guint32 ready_time_secs_high;
+  guint32 ready_time_secs_low;
 } EpgFakeSource;
 
 static gboolean
 epg_fake_source_check (GSource *source)
 {
+  gint64 ready_time_secs;
   EpgFakeSource *self = (EpgFakeSource *) source;
 
-  return epg_clock_get_time (EPG_CLOCK (self->clock)) >= self->ready_time_secs;
+  ready_time_secs = self->ready_time_secs_high;
+  ready_time_secs = (ready_time_secs << 32) +
```

Comment:
This is not quite correct if `self->ready_time_secs_high == G_MAXUINT32` right? Since won't `ready_time_secs` use one bit to store its sign?

And could we do it in one line and use bitwise OR: `ready_time_secs = (gint64) (self->ready_time_secs_high << 32) | self->ready_time_secs_low`

Reply:
I don't see the problem - ready_time_secs is indeed 63-bits of number and 1-bit of sign, but bit shifting left doesn't care about signs at all, and while bit shifting right sign extends, I don't see how that can hurt us.  Am I missing something?

In your other suggestion you would need to cast inside the braces to avoid incorrect computation (and a compiler warning).

| and + should give exactly the same answer here because _low is unsigned and being combined with nothing but 0s.

Is there really a concern about the number of lines of code this takes? I find it more obvious what's happening without having to deal with explicit typecasting.

Reply:
I was worried that If `ready_time_secs` were somehow negative, then `ready_time_secs_high = ready_time_secs >> 32;` would not set the leftmost bit of `ready_time_secs_high` if the compiler noticed that it's assigning to an unsigned integer. But experimentally from running [this](https://gist.github.com/mwleeds/0437d23a6c28d47c1d6e3e8a77bf39b2), it appears this does work.

Reply:
Never mind about the one-line solution. Let's keep what you have, especially after I've tested it independently.